### PR TITLE
creds: Add ImportCredential=

### PR DESCRIPTION
**Pull request by daandemeyer**
ImportCredential= takes a credential name and searches for a matching
credential in all the credential stores we know about it. It supports
globs which are expanded so that all matching credentials are loaded.

### CODE REVIEW

**Comment by poettering** [comment]
hmm, so adding globbing to the existing LoadCredentials= is hard, since both "*", "?" are currently allowed in credential names, so this would create ambiguity/compat problem.

so here's another problem that is currently not addressed: we currently place LoadCredential= into all our well known services for various creds, but we never also add LoadEncryptdCredential= next to it, thus the EFI encrypted TPM import is not actually hooked up automatically right now on all those services. Now we could add both LoadCredential= and LoadEncryptedCredential= to all those services, but that feels a bit "redundant" in a way.

So maybe, if we have to add something new anyway, maybe let's add a single new setting ImportCredential=. It would not be able to import arbitrary files (which is what LoadCredential= and LoadEncryptedCredential= would be really about in the future), but it would be only for importing creds from the service manager. And it would support globbing, out of the box, and automatically look into the encrypted dirs, and decrypt then. We'd then change all our default units to use that.

or in other words:

```ImportCredential=waldo.*``` would glob through both the encrypted and the decrypted system creds dirs, and then decrypt if necessary.

**Comment by poettering** [comment]
hmm, alternatively, we could call it PassCredential= to match the existing PassEnvironment= which would is very similar, just for env vars instead of creds...

**Comment by poettering** [comment]
ah, fuck, we probably shouldn't call it "PassCredential=" because that would be utterly confusing, because we expose "PassCredentials=" already for .socket units, where it controls SO_PASSCRED which is something totally and utterly different.

File: src/core/execute.c [hunk 1]
```diff
@@ -2613,24 +2613,24 @@ static char **credential_search_path(
 
         assert(params);
```
Comment (poettering) [inline src/core/execute.c:2775]: so this commit I am not sure about. The original behaviour here is actually precisely what I had in mind: if you use LoadEncryptedCredential= we'd definitely decrypt and authenticate (in particular the latter) what we find, regardless where we find it.

i.e. the idea was that if LoadEncryptedCredential= is used, there's really no implicit trust: we strictly validate at time of load, and refuse if not valid. 

That's why I wanted to have ImportCredential= with friendlier semantics: so that there we'll decrypt/auth if necessary, but not if if not. It would not be as strict: it would trust any previous implied trust.

File: src/shared/bus-unit-util.c [hunk 2]
```diff
@@ -1200,6 +1200,37 @@ static int bus_append_execute_property(sd_bus_message *m, const char *field, con
...
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
+                r = sd_bus_message_close_container(m);
+                if (r < 0)
+                        return bus_log_create_error(r);
+
```
Comment (poettering) [inline src/shared/bus-unit-util.c:1219]:
```c
if (isempty(eq))
        r = sd_bus_message_append(m, "(sv)", field, "as", 0);
else
        r = sd_bus_message_append(m, "(sv)", field, "as", 1, eq);
if (r < 0)
        …
```

File: src/core/load-fragment.c [hunk 3]
```diff
@@ -4861,35 +4900,64 @@ int config_parse_load_credential(
...
+                log_syntax(unit, LOG_WARNING, filename, line, r, "Failed to resolve unit specifiers in \"%s\", ignoring: %m", s);
+                return 0;
+        }
+        if (!filename_is_valid(s)) {
+                log_syntax(unit, LOG_WARNING, filename, line, 0, "Credential name \"%s\" not valid, ignoring.", s);
+                return 0;
+        }
+
+        r = set_put_strdup(import_credentials, s);
+        if (r == -EEXIST) {
```
Comment (keszybz) [inline src/core/load-fragment.c]: Hmm, `set_put_strdup()` returns 0 if the entry exists, so this would never trigger.
But looking at it more closely, I don't think it's useful to warn. It hardly seems an error if the same glob or name is specified twice. So maybe just make this `LOG_DEBUG`?

File: test/units/testsuite-54.sh [hunk 4]
```diff
@@ -104,6 +104,21 @@ systemd-run -p LoadCredential=cred:/tmp/ts54-creds \
...
+mkdir -p /etc/credstore
+echo -n c >/etc/credstore/test.creds.third
+systemd-run -p "ImportCredential=test.creds.*" \
+            -p DynamicUser=1 \
+            --wait \
+            --pipe \
+            cat '${CREDENTIALS_DIRECTORY}/test.creds.first' \
+                '${CREDENTIALS_DIRECTORY}/test.creds.second' \
+                '${CREDENTIALS_DIRECTORY}/test.creds.third'>/tmp/ts54-concat
+( echo -n abc ) | cmp /tmp/ts54-concat
```
Comment (keszybz) [inline test/units/testsuite-54.sh]: Add space before `>`.

Why is `echo` wrapped in a subshell? Maybe just
```
cmp /tmp/ts54-concat <(echo -n 'abc')
```

File: units/systemd-firstboot.service [hunk 5]
```diff
@@ -29,10 +29,7 @@ StandardError=tty
...
 # manager. This is useful for importing this data from nspawn's
 # --set-credential= switch.
-LoadCredential=passwd.hashed-password.root
-LoadCredential=passwd.plaintext-password.root
-LoadCredential=passwd.shell.root
-LoadCredential=firstboot.locale
-LoadCredential=firstboot.locale-messages
-LoadCredential=firstboot.keymap
-LoadCredential=firstboot.timezone
+ImportCredential=passwd.hashed-password.root
```
Comment (keszybz) [inline units/systemd-firstboot.service:35]: I think the changes to unit files would better be a separate commit.

**Review by keszybz — CHANGES_REQUESTED** [review]
LGTM. The docs have some preexisting issues that I noticed when looking at your changes, but I think we can leave them for later. They are not directly related to the changes here.

**Comment by DaanDeMeyer** [comment]
Implemented the suggestions, setting the green label

**Review by bluca — APPROVED** [review]

File: src/core/execute.c [hunk 6]
```diff
@@ -2802,6 +2803,111 @@ static char **credential_search_path(const ExecParameters *params, CredentialSea
...
+                if (!j)
+                        return -ENOMEM;
+
+                r = safe_glob(j, 0, &pglob);
+                if (r == -ENOENT)
+                        continue;
+                if (r < 0)
+                        return r;
+
+                for (unsigned n = 0; n < pglob.gl_pathc; n++) {
```
Comment (poettering) [inline src/core/execute.c:2870]: gl_pathc is of type size_t, hence `n` should be

Comment (poettering) [reply src/core/execute.c:2870]: → https://github.com/systemd/systemd/pull/27991
